### PR TITLE
[WIP] For #19947: Adding new logins

### PR DESCRIPTION
**Pull request by VitalyVPinchuk**
For #19947 
WIP

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt [hunk 1]
```diff
@@ -56,6 +57,48 @@ open class SavedLoginsStorageController(
         }
     }
 
+    fun add(hostnameText: String, usernameText: String, passwordText: String) {
```
Comment (eliserichards) [inline app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt:60]: I think this function should be `suspend`, and the other one shouldn't. That will simplify the `launch` here

File: app/src/main/java/org/mozilla/fenix/settings/logins/fragment/AddNewLoginFragment.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import org.mozilla.fenix.ext.toEditable
+import org.mozilla.fenix.settings.logins.LoginsFragmentStore
+import org.mozilla.fenix.settings.logins.controller.SavedLoginsStorageController
+import org.mozilla.fenix.settings.logins.createInitialLoginsListState
+import org.mozilla.fenix.settings.logins.interactor.EditLoginInteractor
+
+/**
+ * Displays the editable new login information for a single website
+ */
+class AddNewLoginFragment : Fragment(R.layout.fragment_add_new_login) {
```
Comment (eliserichards) [inline app/src/main/java/org/mozilla/fenix/settings/logins/fragment/AddNewLoginFragment.kt:31]: I wonder if there's a naming convention we should follow here. Take a look to see if credit cards have "add new" yet and see what they named it

File: app/src/main/java/org/mozilla/fenix/settings/logins/fragment/AddNewLoginFragment.kt [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        binding.usernameText.text = "".toEditable()
+        binding.passwordText.text = "".toEditable()
+
+        loginsFragmentStore = StoreProvider.get(this) {
+            LoginsFragmentStore(
+                createInitialLoginsListState(requireContext().settings())
+            )
+        }
+
+        interactor = EditLoginInteractor(
```
Comment (eliserichards) [inline app/src/main/java/org/mozilla/fenix/settings/logins/fragment/AddNewLoginFragment.kt:56]: Nit: we'll want our own interactor for this class

File: app/src/main/res/layout/fragment_add_new_login.xml [hunk 4]
```diff
@@ -0,0 +1,253 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- This Source Code Form is subject to the terms of the Mozilla Public
+   - License, v. 2.0. If a copy of the MPL was not distributed with this
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+
+<androidx.constraintlayout.widget.ConstraintLayout
```
Comment (eliserichards) [inline app/src/main/res/layout/fragment_add_new_login.xml:6]: This is pretty much identical to the "edit" fragment, right?

File: app/src/main/res/layout/fragment_add_new_login.xml [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            android:singleLine="true"
+            android:textColor="?primaryText"
+            android:textCursorDrawable="@null"
+            android:textSize="16sp"
+            android:textStyle="normal"
+            app:backgroundTint="?primaryText"
+            tools:ignore="Autofill" />
+    </com.google.android.material.textfield.TextInputLayout>
+
+    <ImageButton
```
Comment (eliserichards) [inline app/src/main/res/layout/fragment_add_new_login.xml:204]: I think we can remove this button here. I don't think we need it if we're creating a new item

**Review by eliserichards — COMMENTED** [review]
Looking good so far!

File: app/src/main/res/layout/fragment_add_new_login.xml (hunk 4 above)
Comment (VitalyVPinchuk) [reply app/src/main/res/layout/fragment_add_new_login.xml:6]: Yes, it is. Except for hostname text is clickable and focusable.

File: app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt (hunk 1 above)
Comment (VitalyVPinchuk) [reply app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt:60]: I made it similar to `fun save()`
Should I still make it the other way?

File: app/src/main/java/org/mozilla/fenix/settings/logins/fragment/AddNewLoginFragment.kt (hunk 2 above)
Comment (VitalyVPinchuk) [reply app/src/main/java/org/mozilla/fenix/settings/logins/fragment/AddNewLoginFragment.kt:31]: Sure!
Well, there's no special class for adding new card but it's the same for editing and adding.
There's no "new" word, just `onAddCreditCardClick()` or `handleAddCreditCardClicked()`
Should it sound similar way?

File: app/src/main/java/org/mozilla/fenix/settings/logins/fragment/AddNewLoginFragment.kt (hunk 3 above)
Comment (VitalyVPinchuk) [reply app/src/main/java/org/mozilla/fenix/settings/logins/fragment/AddNewLoginFragment.kt]: Fixed

**Comment by st3fan** [comment]
@VitalyVPinchuk thank you for picking this up!

File: app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt (hunk 1 above)
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt:60]: let me take a look. I guess keeping it the same way as the other stuff is fine

File: app/src/main/java/org/mozilla/fenix/settings/logins/interactor/EditLoginInteractor.kt [hunk 6]
```diff
@@ -21,4 +21,8 @@ class EditLoginInteractor(
     fun onSaveLogin(loginId: String, usernameText: String, passwordText: String) {
         savedLoginsController.save(loginId, usernameText, passwordText)
     }
+
+    fun onAddNewLogin(hostnameText: String, usernameText: String, passwordText: String) {
```
Comment (eliserichards) [inline app/src/main/java/org/mozilla/fenix/settings/logins/interactor/EditLoginInteractor.kt:25]: Nit: 
```suggestion
    fun onAddLogin(hostnameText: String, usernameText: String, passwordText: String) {
```

File: app/src/main/res/layout/fragment_add_new_login.xml [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+<!--    <com.google.android.material.button.MaterialButton-->
+<!--        android:id="@+id/save_button"-->
+<!--        style="@style/NeutralButton"-->
+<!--        android:layout_width="wrap_content"-->
+<!--        android:paddingStart="12dp"-->
+<!--        android:paddingEnd="12dp"-->
+<!--        android:text="@string/new_login_save_button"-->
+<!--        app:layout_constraintBottom_toBottomOf="parent"-->
+<!--        app:layout_constraintEnd_toEndOf="parent"-->
+<!--        app:layout_constraintTop_toBottomOf="@+id/inputLayoutPassword" />-->
```
Comment (eliserichards) [inline app/src/main/res/layout/fragment_add_new_login.xml]: This can be removed now 👍

File: app/src/main/res/navigation/nav_graph.xml [hunk 8]
```diff
@@ -344,6 +344,11 @@
         <action
             android:id="@+id/action_savedLoginsFragment_to_loginDetailFragment"
             app:destination="@id/loginDetailFragment" />
+        <action
+            android:id="@+id/action_savedLoginsFragment_to_addNewLoginFragment"
```
Comment (eliserichards) [inline app/src/main/res/navigation/nav_graph.xml:348]: Nit: 
```suggestion
            android:id="@+id/action_savedLoginsFragment_to_addLoginFragment"
```

File: app/src/main/res/navigation/nav_graph.xml [hunk 9]
```diff
@@ -388,6 +393,18 @@
             app:popUpToInclusive="true" />
     </fragment>
 
+    <fragment
+        android:id="@+id/addNewLoginFragment"
```
Comment (eliserichards) [inline app/src/main/res/navigation/nav_graph.xml:397]: Nit:
```suggestion
        android:id="@+id/addLoginFragment"
```

File: app/src/main/res/navigation/nav_graph.xml [hunk 10]
```diff
@@ -388,6 +393,18 @@
...
+        android:name="org.mozilla.fenix.settings.logins.fragment.AddNewLoginFragment"
+        android:label="@string/add_login"
+        tools:layout="@layout/fragment_add_new_login">
+        <action
+            android:id="@+id/action_addNewLoginFragment_to_savedLoginsFragment"
```
Comment (eliserichards) [inline app/src/main/res/navigation/nav_graph.xml:402]: same... remove "new" in all of these :)

**Review by eliserichards — CHANGES_REQUESTED** [review]

**Comment by eliserichards** [comment]
@Mergifyio rebase

**Comment by eliserichards** [comment]
Screenshots: 

image  image

**Review by eliserichards — APPROVED** [review]
Awesome job! I added some tests/cleanup and will run the CI in https://github.com/mozilla-mobile/fenix/pull/21199 🎉 🙌 🚀

**Pull request closed without merging**